---
title: "Lab2 Preprocessing"
author: "ks"
date: "May 23, 2018"
output:
  html_document: default
  pdf_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install libraries

The following are the libraries we will be using. Let's install them if they are not already installed.
```{r libraries}
source("http://bioconductor.org/biocLite.R") 
if(!require("genefilter")) {biocLite("genefilter")}
if(!require("vsn")) {biocLite("vsn")}
if(!require("latticeExtra")) {install.packages("latticeExtra")}
if(!require("sva")) {biocLite("sva")}
if(!require("ruv")){biocLite("ruv")}
```

## Import Data from GenomeStudio Report files

Two summary report files were output from GenomeStudio:

1. irina-spp.txt contains the sample (biological) feature data and 
2. irina-cpp.txt contains the control feature data

These files contain the variables: Avg_Signal, Detection p, Avg_NBEADS, and BEAD_STDERR, for the Prostate Cancer cell line data (Ianculescu et al. J Biol Chem., 2012) (GSE31873).

Note: Tips if you are ever outputting data from GenomeStudio:

 1. add additional data columns (hidden by default);
 2. increase number of significant figures;
 3. output imputed probes.

```{r read data}
library(limma)
datadir=c("data/JBC 2012") 
sfile<-file.path(datadir,"irina-spp.txt") #standard probe profile output from GenomeStudio
cfile<-file.path(datadir,"irina-cpp.txt") #control probe profile output from GenomeStudio
rawdata = read.ilmn(files=sfile, ctrlfiles=cfile,
        other.columns=c("Detection","Avg_NBEADS","BEAD_STDERR"))
targets=read.csv(file=file.path(datadir,"SampleChars.csv"),row.names=1) #treatment information
rawdata$targets=targets[colnames(rawdata),] # create slot for sample characteristic data in list object housing the gene expression data order rows to match columns of gene expression (arrays) 
rm(targets)
```
```{r data list}
names(rawdata)
rawdata$source
```


```{r dimE}
dim(rawdata$E)
```
```{r list cpp}
head(rawdata$genes,n=10)
head(unique(rawdata$gene$Status),n=10)
```
```{r table cpp}
rawdata$gene$Status[which(substr(rawdata$gene$Status,1,4)=="ERCC")]="ERCC"
table(rawdata$gene$Status)
```
```{r targets}
head(rawdata$targets,n=3) 
```
```{r design}
table(rawdata$targets$Type)
```


## Make boxplot of log2 intensities 

```{r negctls, echo=FALSE}
boxplot(log2(rawdata$E[rawdata$genes$Status=="NEGATIVE",]),range=0,    xlab="Arrays",ylab="log2 intensities",main="Negative control probes")
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r sampleFeatures, echo=FALSE}
boxplot(log2(rawdata$E[rawdata$genes$Status=="regular",]),range=0,
    xlab="Arrays",ylab="log2 intensities",main="Regular probes")
```

We see variation in distribution of intensities across arrays. Some variation is technical noise, some is signal.
This similar distributions of features across arrays makes quantile normalization a possibility.

## Make density plot of log2 intensities 

Make density plot of 'regular' probes. Plot 1 sample first, then add other samples as lines to plot
```{r density, echo=FALSE}
plot(density(log2(rawdata$E[rawdata$genes$Status=="regular",1])),
	xlab="log2 intensities", ylab="Density",
	main="Density Plot of Raw Intensities",ylim=c(0,1.3)) 
  na=length(colnames(rawdata$E))
  for (i in 2:na)
	    lines(density(log2(rawdata$E[rawdata$genes$Status=="regular",i])),col=i,lty=i)
legend(12,1.2,colnames(rawdata$E),lty=1:5,col=1:6,cex=.9) 
```

This figure supports the previous one, showing random variation but no noticeable outlier (failed) samples.

## Make boxplots of the number of beads/probe 

The number of beads per probe is centered around 20.
```{r nbeads, echo=FALSE}
par(mfrow=c(1,2))
boxplot(rawdata$other$Avg_NBEADS[rawdata$genes$Status=="regular",],
    main="Regular probes",xlab="Arrays",ylab="Number of Beads") 
boxplot(rawdata$other$Avg_NBEADS[rawdata$genes$Status=="NEGATIVE",],
    main="Negative Control Probes",xlab="Arrays",ylab="Number of Beads") 
```

NOTE: these figures will only display if the exported GenomeStudio
Files report Avg_NBEADS in 'other.columns'.

## MA plots: log-fold change vs average log expression 

Let's look at MA plots of log fold-change vs. average log expression. We will compute a synthetic array using the probewise medians, and compare each individual array to the synthetic array. We are looking for arrays that have oscillating lowess curves.
```{r maplots, echo=FALSE}
mdArray=apply(log2(rawdata$E[rawdata$genes$Status=="regular",]),1,median)
par(mfrow=c(2,3))
for (i in 1:6) {
M<-log2(rawdata$E[rawdata$genes$Status=="regular",i])-mdArray
A<-(log2(rawdata$E[rawdata$genes$Status=="regular",i])+mdArray)/2
smoothScatter(A,M)
lines(lowess(A,M),col=2)
}
```

## Cluster between-sample distances

Look for outlier sample using rawdata. Why before processing?
```{r outlier detection, echo=FALSE}
library(genefilter)
library(latticeExtra)
dd=dist2(log2(rawdata$E[rawdata$genes$Status=="regular",])) #default calculates mean of absolute value of (pairwise) differences
dd.row=as.dendrogram(hclust(as.dist(dd)))
row.ord=order.dendrogram(dd.row)
legend=list(top=list(fun=dendrogramGrob,
                      args=list(x=dd.row,side="top")))
lp=levelplot(dd[row.ord,row.ord],
             scales=list(x=list(rot=90)),xlab="",
             ylab="",legend=legend)
lp
rm(dd,row.ord,lp)
```

I do not see any sample that stands out especially and needs removal.

##Estimate the proportion of Genes Expressed on each array
```{r PropExprGenes}
library(limma)
proportion=propexpr(rawdata)
names(proportion)=rawdata$targets$Type
tapply(proportion,rawdata$targets$Type,mean) 
```
We see similar numbers of expressed genes across samples.  

## Process the raw signals

Let's do the signal processing of the data to remove technical variation. neqc() performs normal-exponential background correction, quantile normalization and log2 transformation.
```{r process}
  jbc=neqc(rawdata)
  dim(jbc) 
```
NOTE: control probes are now removed from the expression set.


Now let's see how tightly the treatment replicates cluster.
```{r mds}
plotMDS(jbc,labels=paste(jbc$targets$Type,unclass(jbc$targets$batch), sep="_"),col=unclass(jbc$targets$Type),xlim = c(-1.5,1.5), ylim=c(-1,1), ndim=2) #color by type
```

Batches (2,4) don't seem to cluster with (1,3)

## Apply ComBat for batch correction

Let's try to correct for batch effects using ComBat.
```{r combat}
library(sva)
batch = unclass(jbc$targets$batch)
design=model.matrix(~factor(jbc$target$Type))
jbc.bc = ComBat(dat=jbc$E, batch=batch, mod=design)
mds<-plotMDS(jbc.bc,labels=paste(jbc$targets$Type,unclass(jbc$targets$batch), sep="_"),col=unclass(jbc$targets$Type),xlim = c(-1.5,1.5), ylim=c(-1,1), ndim=3) #color by type
```
```{r combatMDS, echo=FALSE}
plotMDS(mds,labels=paste(jbc$targets$Type, 
      unclass(jbc$targets$batch), sep="_"),
      col=unclass(jbc$targets$Type),xlim = c(-1.5,1.5), ylim=c(-1,1),
      dim.plot=c(2,3)) #color by type
```

That cleaned up the data so that the 4 replicate treatment effects now cluster tightly with each other.

The ComBat corrected data are the ones stored on GEO.
```{r geoquery}
library(GEOquery)
#gse31873=getGEO('GSE31873',GSEMatrix=TRUE)
load(paste(datadir,"/gse31873.rda",sep="")) #saved locally to load more quickly
show(gse31873)
```
```{r GeoExprs}
c42b=gse31873$GSE31873_series_matrix.txt.gz
ec42b=exprs(c42b)    # exprs() accesses the gene expression values
ec42b[1:4,1:3] 
```
Compare this to the combat-corrected data above:
```{r GeoExprs jbc}
jbc.bc[rownames(ec42b),][1:4,1:3]
```

## Apply RUV-4 for batch correction

Remove Unwanted Variation (RUV) is a method for removing batch effects when the batches are not recorded.

```{r ruv}
library(ruv)
design=model.matrix(~factor(jbc$target$Type))
fit=lmFit(jbc,design)
efit=eBayes(fit)
# This is an arbitrary definition to get ~11,000 features 
# that measure noise (large p)   (~25% of features on array)
enc=efit$p.value[,2]>0.3 & efit$p.value[,3]>0.3 & efit$p.value[,4]>0.3 &
  efit$p.value[,5]>0.3 & efit$p.value[,6]>0.3
table(enc)
myX=matrix(design,ncol=6)[,-1]
ruvfit10=RUV4(Y=t(jbc$E),X=as.matrix(myX),ctl=enc,10)
str(ruvfit10)
# The W matrix provides the estimated covariates that measure unwanted variation to include in analysis. Let's do a regression of Expression on W and do an MDS plot of the residuals.
modW=model.matrix(~ruvfit10$W)
fit=lmFit(jbc$E,modW)
yhat=fit$coef %*% t(modW)
jbc.Wresid=jbc$E-yhat
mds<-plotMDS(jbc.Wresid,labels=paste(jbc$targets$Type,   unclass(jbc$targets$batch), sep="_"),
        col=unclass(jbc$targets$Type),xlim=c(-1.5,1.5),
        ylim=c(-1,1.3),ndim=3) #color by type
```
```{r plotMDS,echo=FALSE}
plotMDS(mds,labels=paste(jbc$targets$Type, unclass(jbc$targets$batch), sep="_"),
             col=unclass(jbc$targets$Type),xlim=c(-1.5,1.5),
             ylim=c(-1,1.3),dim.plot=c(2,3)) #color by type
```

These look sharp. The samples cluster just like they did using the measured batch correction method, ComBat.

## Apply SVA for batch correction

Surrogate Variable Analysis (SVA) is an older method for batch correction without a batch variable.
```{r sva}
design=model.matrix(~factor(jbc$target$Type))
design0= model.matrix(~1,data=as.data.frame(jbc$target$Type))
# Now we must choose the number of surrogate variables to estimate.
# the two methods provided by the software often give different results
# I do not have a way of choosing this so I will typically start with 2  
# and see if the results make sense
svobj = sva(jbc$E,design,design0,n.sv=2)
```
```{r svs}
head(svobj$sv)
```
```{r dim-svs}
dim(svobj$sv)
```

Now I will perform regression of expression data on the estimated surrogate variables and save residuals for plotting
```{r sva plot}
modsv=model.matrix(~svobj$sv)
fit=lmFit(jbc$E,modsv)
yhat=fit$coef %*% t(modsv)
jbc.svresid=jbc$E-yhat
mds<-plotMDS(jbc.svresid,labels=paste(jbc$targets$Type, unclass(jbc$targets$batch), sep="_"),
             col=unclass(jbc$targets$Type),xlim=c(-1.5,1.5),
             ylim=c(-1,1.3),ndim=3) #color by type (slide 47)
```
```{r sva plot2, echo=FALSE}
plotMDS(mds,labels=paste(jbc$targets$Type, unclass(jbc$targets$batch), sep="_"),
        col=unclass(jbc$targets$Type),xlim=c(-1.5,1.5),
        ylim=c(-1,1.3),dim.plot=c(2,3))
```


```{r sessionInfo}
sessionInfo()
```