---
title: "Lab3 Differential Expression"
author: "ks"
date: "5/29/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(limma)
```

Let's analyze the prostate cancer cell line data we've used in class. The treatments are:  

treatment 1: siNS, sip300, siCBP 

treatment 2:  0 hr,  16 hrs
     
All cells are treated with androgen at 0 hrs. The genes that change expression between 0 and 16 hours are genes that are regulated by androgen. These might vary based on the different treatments siNS (control), sip300, and siCBP. Let's compare average gene expression between 0 and 16 hrs.

### 1. Load the data (from JBC (2012)).
```{r ReadData}
jbcdir=c("data/JBC 2012")
load(file.path(jbcdir,"jbcdat.rda"))
```

### 2. Create a plot of fold-change vs average (log) intensity

```{r MAplot, echo=FALSE}
Index1 <- which(jbcdat$target$hour=="0hr")
Index2 <- which(jbcdat$target$hour=="16hr")

d <- rowMeans(jbcdat$E[,Index2])-rowMeans(jbcdat$E[,Index1])
a <- rowMeans(jbcdat$E)

par(mfrow=c(1,2))
plot(a,d,xlab="Average (log) intensity",ylab="Average log fold-change",main="MA-plot",pch=".")
smoothScatter(a,d,nrpoints=500,
              xlab="Average (log) intensity",
              ylab="Average log fold-change",
              main="MA plot")
abline(-1,0,col=7,lwd=3)
abline(1,0,col=7,lwd=3)
abline(0,0,col=2,lwd=3)
```

Why might one prefer the figure on the right?

### 3. 2-sample t-tests for each gene
```{r ttests}
tt <- genefilter::rowttests(jbcdat$E,factor(jbcdat$targets$hour))
head(tt,n=3)
```

These are equivalent to the Ftest with equal variances: 

```{r Ftests}
Ft <- genefilter::rowFtests(jbcdat$E,factor(jbcdat$target$hour),var.equal=TRUE)
head(Ft,n=3)
```

Now let's display the results using a Volcano Plot of -log10 pvalue vs (log) fold-change
```{r volcanoPlot,echo=FALSE}
tt <- genefilter::rowttests(jbcdat$E,factor(jbcdat$targets$hour))

lodt <- -log10(tt$p.value)
plot(d,lodt,main="Volcano plot for t-test",xlab="Average fold-change",
	ylab="-log10 pvalue",pch=".")
# make larger data points when -log10 pvalue > 10 
#  pch = 20 gives solid dots, col=4 is blue
points(d[lodt>10],lodt[lodt>10],pch=20,col=4)
```

We can give emphasis to highly sigificant pvalues (-log10p > 10).

And again, smooth scatter can be used.

```{r smoothedVolcano,echo=FALSE}
smoothScatter(d,lodt,nrpoints=500,
              xlab="Average fold-change",
              ylab="-log10 pvalue",
              main="Volcano plot for t-test")
# If you want to highlight high lodt & fold change
# col= 6 is pink
points(d[lodt>10 & abs(d)>1],lodt[lodt>10 & abs(d)>1],pch=20,col=6)
```

### 4. 2-sample moderated t-tests for each gene

Now let's see how the results change if we run a moderated t-test.
```{r modT}
design <- model.matrix(~factor(jbcdat$targets$hour))

fit  <- limma::lmFit(jbcdat$E,design)
efit <- limma::eBayes(fit)
```

lab3 Supplemental code.R will show how the moderated t-test and pvalue are computed.

```{r modTvolcanoplot, echo=FALSE}
lodmt <- -log10(efit$p.value[,2])
smoothScatter(efit$coef[,2],
              lodmt,
              nrpoints=500,
              main="Volcano plot for moderated t-test",xlab="Average fold-change",
	ylab="-log10 pvalue")
```

The limma package has a function for this.  However, it doesn't use scatterSmooth.

```{r volcanoplot, echo=FALSE}
volcanoplot(efit, coef = 2, style = "p-value", highlight = 10, names = jbcdat$genes$Symbol, hl.col="blue",
            xlab = "Log2 Fold Change", ylab = NULL, pch=16, cex=0.35)
```

Let's compare the pvalues from student's t-test and the moderated t-test.
```{r ttvmodt}
par(mar=c(5,5,3,2))
plot(lodt,lodmt,pch=".",cex.axis=1.2,cex.lab=1.2,
	xlab="-log10(p), 2-sample t-test",
	ylab="-log10(p), moderated t-test")
abline(0,1,col=2,lwd=3)
```

Did the difference in group means change?

### * Table of t-test results

The command topTable will summarize the regression results, ranked by (decreasing) p-value.

```{r TopT}
topTable(efit,coef=2,n=4)
#?topTable
```

### 5. Moderated testing when there are 3 groups

Now let's look for treatment effects at time 16 hours. 

```{r Trt@16}
time <- jbcdat$targets$hour
trts <- factor(jbcdat$targets$treatments, 
            levels=c("siNS","siCBP","sip300"))
design <- model.matrix(~trts)
design[1:4,]
fit <- lmFit(jbcdat$E,design,subset=c(time=="16hr"))

contr.matrix <- cbind(siCBP=c(0,1,0),sip300=c(0,0,1))
contr.matrix
fitgpd <- contrasts.fit(fit,contr.matrix)
fitgpd <- eBayes(fitgpd)
```

```{r topTablecontr}
tt <- topTable(fitgpd,n=5)
tt
# fitgpd$coef[rownames(tt),]
```

If we find differential expression by treatment, what does it mean?

### * Venn diagram summarizing significant results

```{r Venn2}
results <- decideTests(fitgpd,
                       method = "separate", 
                       adjust.method="none",
                       p.value = 0.05,
                       lfc = 0)
vennDiagram(results)
```

Here is the table of counts:
```{r results}
table(results[,c("sip300")],results[,c("siCBP")])
```

Both significant: 2648 = 1095 + 1277 + 167 + 109 

Only siCBP significant: 2733 = 1494 + 1239 

Now we can also break these down to include the information on direction of effect, up-regulation or down-regulation.

```{r VennUpDown}
vennDiagram(results,include=c("up","down"),
            counts.col=c("red", "blue"))
```

The counts for up-regulated genes only:
```{r countsUp}
vennCounts(results,include=c("up"))
```

Down-regulated genes only:
```{r countsDown}
vennCounts(results,include=c("down"))
```

### * Change Regression Model Matrix

If we want to study all pairwise comparisons, we can do this by using a different design matrix and contrast matrix.

```{r Trt@162}
design <- model.matrix(~trts-1)
design[1:4,]

fit <- lmFit(jbcdat$E,design,subset=c(time=="16h"))

contr.matrix <- cbind('siCBP-siNS'=c(-1,1,0),
                   'sip300-siNS'=c(-1,0,1),
                   'siCBP-sip300'=c(0,1,-1))
contr.matrix

fitgpd <- contrasts.fit(fit,contr.matrix)
fitgpd <- eBayes(fitgpd)
```

```{r topTableTrt}
topTable(fitgpd,n=5)
```

How is this model the same as the previous model?

How is it different?

```{r Venn}
results <- decideTests(fitgpd,
                       method = "separate", 
                       adjust.method="none",
                       p.value = 0.05,
                       lfc = 0)
vennDiagram(results)
```

### * Saturated (Full) Model

Now look for genes associated with Time, Co-Reg or Time x Co-Reg interaction.
```{r mvm}
time <- jbcdat$targets$hour
trt  <- factor(jbcdat$targets$treatments, 
            levels=c("siNS","siCBP","sip300"))
design <- model.matrix(~trts*time)
#design
fit <- lmFit(jbcdat$E,design)

contr.matrix <- cbind(siCBP=c(0,1,0,0,0,0),sip300=c(0,0,1,0,0,0),Time=c(0,0,0,1,0,0),siCBPxTime=c(0,0,0,0,1,0), sip300xTime=c(0,0,0,0,0,1))
#contr.matrix
fitfull <- contrasts.fit(fit,contr.matrix)
fitfull <- eBayes(fitfull)
topTable(fitfull,n=5)
```

Let's perform a global test of Ho: no difference across any level of time x treatment vs Ha: at least 1 difference.
```{r Ftest}
genesign <- which(fitfull$F.p.value<0.05)
beta.p <- fitfull$p.value[genesign,]
statsign <- ifelse(beta.p<0.05,1,0)
vennDiagram(statsign)
```

### * Plot some individual genes

```{r Type}
plot(1:24,jbcdat$E["ILMN_1726114",order(jbcdat$targets$type)],col=jbcdat$targets$type[order(jbcdat$targets$type)], pch=15,xlab="sample",ylab="log2 E", main="ILMN_1726114")
legend(1,12,levels(jbcdat$targets$type),cex=1.2,pch=15, col=1:6)
```

This gene is differentially expressed with time, and similar for all three treatment groups.

And now ILMN_1759676:
```{r gene3}
plot(1:24,jbcdat$E["ILMN_1759676",order(jbcdat$targets$type)],col=jbcdat$targets$type[order(jbcdat$targets$type)],pch=15,xlab="sample",ylab="log2 E",main="ILMN_1759676",xlim=c(1,30))
legend("topright",levels(jbcdat$targets$type),cex=1.2,pch=15,
       col=1:6)
```

```{r genesymbols}
jbcdat$genes["ILMN_1759676",]
```

The time effect for p300 knockdown is smaller than for the other treatment groups. p300 knockdown has lower expression of HOXC13 at baseline (0h) than the other treatment groups (siNS, siCBP).

### 6. SessionInfo

```{r sessionInfo}
sessionInfo()
```
