---
title: "Lab 5. RNA seq"
author: "ks"
date: "6/4/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Example Data

The data are for 4 experiments in A549 cells downloaded from ENCODE  (Reddy et al. 2009).
```{r loadData}
load(file="data/RNAseq/cnts.rda")
show(cnts)
```

These data are stored as a RangedSummarizedExperiment object. This object was created by the transcript/gene counting function. We'll need some new libraries to reformat this for performing differential expression.
```{r libs}
if(!require("GenomicAlignments")) {BiocManager::install("GenomicAlignments")}
if(!require("edgeR")) {BiocManager::install("edgeR")}
library(GenomicAlignments)
library(edgeR)
library(limma)
```

Let's load the table of counts. 
```{r TableOfCounts}
toc <- assays(cnts)$counts
head(toc)
```

The row names are the Entrez Gene Id numbers.

Let's find the counts for the 2 genes from Reddy et al. that showed differential expression for this experiment. We know the gene Symbol, but not the Entrez Gene Id. Let's find their Entrez Gene Id.

```{r loadDB}
library(org.Hs.eg.db)
keys <- c("NFKBIA","PER1")
sel <- select(org.Hs.eg.db,keys=keys,keytype="SYMBOL",
              columns="ENTREZID")
show(sel)
```

Now let's see the gene counts for these 2 genes.  In Reddy et al., NHKBIA shows increased expression with DEX treatment.
```{r selectData}
toc[sel$ENTREZID,]
```

If we wanted to know the gene symbols for all the genes in our count matrix, we can link the names in the other direction.
```{r eg2symbol}
keys <- rownames(toc)
sel <- select(org.Hs.eg.db,keys=keys,columns="SYMBOL",
            keytype="ENTREZID")
show(sel[1:5,])
```


Next, we'll get the treatment information from the column names.
```{r trtinfo}
group <- sub(c("Rep."),"",colnames(cnts))
group <- factor(group,levels=c("Etoh02","Dex100nm"))
group
```

The default ordering when you make a factor variable is to alphabetize the groups. Here, since Etoh02 is the control experiment, I want appearing first so I set this using 'levels=c("Etoh02","Dex100nm")'.

## Differential Expression using the limma package

### 1. eBayes, with trend=TRUE

For microarray data analysis, the sd was approximately independent of average intensity, so we assumed a single variance estimate for the prior and shrunk the feature-specific sds to the prior. With log2 count data, we're going to see the relationship between sd and average log2 intensity is not constant, so we model it using a trend. Then we shrink our variance estimates to a point on the trend-line.

We begin with creating a list variable (DGEList) used for differential expression of count data.
```{r dgelist}
dge <- edgeR::DGEList(counts=assays(cnts)$counts,
             group=group)
dim(dge)
```

Show summary of data table.
```{r datatable}
dge$samples
```


```{r headcounts}
head(dge$counts)
```

First step is to remove rows with 0 or consistently low counts.
```{r filter}
 design <- model.matrix(~dge$samples$group)
 keep <- edgeR::filterByExpr(dge, design)
 dge <- dge[keep,,keep.lib.sizes=FALSE]
 dim(dge)
```

Next, we apply a scale-normalization to the read counts, to normalize the samples for variation in library size (total read counts).  We will use the default method, TMM (trimmmed mean of M-values. ...to be discussed another time.)
```{r TMM}
dge <- edgeR::calcNormFactors(dge)
```

Let's look at the distribution of gene counts.

```{r boxplot,echo=F}
boxplot(log2(dge$counts+1),col=as.numeric(group)+1)
```

Do samples cluster by treatment group?
```{r mdsplot, echo=F}
limma::plotMDS(dge,labels =dge$samples$group,lwd=3,cex=1.5,
        cex.axis=1.3,cex.lab=1.5)
box(lwd=3)
```

```{r libsizes}
  dge$samples
```
 
When the library sizes are more or less similar (ratio of largest to smallest is < 3), 'limma-trend' is recommended for differential expression.

First, we transform the counts to counts per million (libsize). A prior count is used to shrink the variances of the logarithms of low counts.

```{r lcpm}
logCPM <- cpm(dge, log=TRUE, prior.count=3)
```

These logCPM values can be used in any standard limma analysis, using the option trend=TRUE when running eBayes to allow the relationship of variance vs intensity to follow a trend.

```{r eBayes}
fit <- limma::lmFit(logCPM, design)
# fit has 2 coefficients: intercept + Coef of interest
contr.matrix <- as.matrix(c(0,1))
colnames(contr.matrix) <- c("Dex100")
# This will reduce the model output to the coef of interest only
fit <- contrasts.fit(fit,contr.matrix)
efit <- limma::eBayes(fit, trend=TRUE)
limma::topTable(efit)
```

How many tests are statistically significant?
```{r decideTests}
summary(de <- limma::decideTests(efit, p=0.05))
```

Now let's show them on an MA plot.
```{r MAplot}
detags <- which(as.logical(de))
limma::plotMD(efit)
points(efit$Amean[detags],efit$coef[detags],
     pch=16,cex=0.7,col="red")
abline(-1,0,col="blue")
abline(1,0,col="blue")
```

### 2. Transform counts using Voom

If the library sizes vary widely across samples, 
tranforming the counts to microarray-like measures using the function voom is recommended.
```{r voom}
design <- model.matrix(~dge$samples$group)
v <- limma::voom(dge,design,plot=TRUE)
```

Now I can do the moderated t-test analysis (using eBayes) on the voom-transformed data.
```{r eBayesv}
fit <- limma::lmFit(v,design)
fit <- contrasts.fit(fit,contr.matrix)
vfit <- limma::eBayes(fit)
```

What does the table of top hits look like?
```{r tophits}
options(digits=3)
limma::topTable(vfit)
```


Let's count the differentially expressed genes with FDR adjusted p<0.05.
```{r decideTestsv}
summary(de <- limma::decideTests(vfit, p=0.05))
```

Now let's show them on an MA plot.
```{r MAplotv}
detags <- which(as.logical(de))
limma::plotMD(vfit)
points(vfit$Amean[detags],vfit$coef[detags],
     pch=16,cex=0.7,col="red")
abline(-1,0,col="blue")
abline(1,0,col="blue")
```

How similar are these results?
```{r CompareResults}
table(trendlimma=limma::decideTests(efit,p=0.05),
      voom=limma::decideTests(vfit,p=0.05))
```


```{r sessionInfo}
sessionInfo()
```
