---
title: "Lab5-ReadCounting"
author: "ks"
date: "6/20/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Genomic annotation (position) for genes

```{r cars}
library(GenomicFeatures)
hg19.tx<-loadDb("~kims/Google Drive File Stream/My Drive/PM579/annotations/hg19.txdb.knownGene")
isActiveSeq(hg19.tx)[seqlevels(hg19.tx)]<-FALSE
isActiveSeq(hg19.tx)[paste("chr",c(1:22,"X","Y"),sep="")]<-TRUE
hg19.tx
```

Now let's create a GRangesList of Exons by Genes, from the UCSC knowngene data base.
```{r knowng}
exByGn <- exonsBy(hg19.tx, "gene")
exByGn
```

Let's plot the exon intervals for the first gene. First, I need to go back and get the plotRanges function we wrote in the earlier tutorial on IRanges.
```{r plotR, echo=FALSE}
plotRanges <- function(x, xlim = x, main = deparse(substitute(x)),
                       col = "black", sep = 0.5, ...)
{
  height <- 1
  if (is(xlim, "Ranges"))
    xlim <- c(min(start(xlim)), max(end(xlim)))
  bins <- disjointBins(IRanges(start(x), end(x) + 1))
  plot.new()
  plot.window(xlim, c(0, max(bins)*(height + sep)))
  ybottom <- bins * (sep + height) - height
  rect(start(x)-0.5, ybottom, end(x)+0.5, ybottom + height, col = col, ...)
  title(main)
  axis(1)
}
```

```{r plotgene1}
  ir1<-ranges(unlist(exByGn[1]))
  plotRanges(ir1)
```

The 5th gene also has overlapping exons. 
```{r plotgene5}
  ir1<-ranges(unlist(exByGn[5]))
  plotRanges(ir1)
```

# ENCODE data:  4 samples (A549 cells)

Now let's read in the data.   Each file is ~1GB and too large to share by Dropbox, so to try and home you will need to download the files from the ENCODE website (see class notes). I'll start by creating a list of *.bam filenames.

```{r data}
datadir=c("~kims/Google Drive File Stream/My Drive/PM579/Data/A549")
fls=dir(datadir,pattern="*\\.bam$")
fls=file.path(datadir,fls)
```

```{r ga}
## reads sorted by qname
library(GenomicAlignments)
bfls <- BamFileList(fls[1:4], index=character(),
                      obeyQname=TRUE, yieldSize=1000000)
bfls
```

##QC functions

The package EDASeq has some nice QC functions.

```{r edaseq, echo=TRUE}
source("http://bioconductor.org/biocLite.R")
if(!require("EDASeq")) {biocLite("EDASeq")}
library(EDASeq)
colors <- c(rep(rgb(1,0,0,alpha=0.7),2),
            rep(rgb(0,0,1,alpha=0.7),2))
barplot(bfls,las=2,col=colors)
```


Note: the line above might take some time to run.  When it's done, we'll see that one of the libraries has many more reads then the others.

The next plot shows the frequency of the different bases by position in the read.
```{r NtFreq}
plotNtFrequency(bfls[[1]])
```

## Gene counting

Now let's count the reads overlapping a gene model.   
We will get the counts over exons (ranges), by gene (feature).  Each read is counted at most one time. These .bam files were from single-end reads and a strand-unaware assay. For paired-end reads include option: single.end=FALSE, ignore.strand=TRUE.
 
```{r count}
cnts <- summarizeOverlaps(exByGn, bfls, mode=Union,ignore.strand=TRUE)
```
This took ~2 min 15 sec on my Mac, and did not finish in 4 hrs on PC.  

The counts are stored in a RangedSummarizedExperiment object.
```{r cnts}
show(cnts)
```

We can access the count table as follows.
```{r head}
head(assays(cnts)$counts)
```

The filenames are used as the column(sample) names. Let's simplify the column names:
```{r simpl}
cn=sub(c("wgEncodeHaibRnaSeqA549"),"",colnames(cnts))
cn=sub(c(".bam"),"",cn)
cn=sub(c("Aln"),"",cn)
colnames(cnts)=cn
#order with controls first (right)
cnts=cnts[,c(3,4,1,2)]
cnts
head(assays(cnts)$counts)
```

Since this object took a long time to create, I'm going to save the object as an R data set for easy loading by all.
```{r save}
#save(cnts,file="~kims/GitHub/PM579/data/RNAseq/cnts.rda")
```

```{r sI}
sessionInfo()
```

