---
title: "Small sample t tests"
author: "ks"
date: "5/29/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if(!require("genefilter")){BiocManager::install("genefilter")} 
library(genefilter)
library(tidyverse)
```

### Scenario:  Test differential expression of 20000 genes 

Suppose we want to test for differential expression in two treatment groups (treatment vs control), each run in triplicate.
Let's run a simulation to see how the t-test behaves in small sample sizes of 3 vs 3, when there is no treatment effect (null hypothesis).  What mean and sd should we use?

I propose getting the mean gene expression profile and estimates of standard deviation from the data set we've been using in class/labs (jbcdat.rda). 

Let's load the data from JBC (2012).
```{r ReadData}
jbcdir=c("data/JBC 2012")
load(file.path(jbcdir,"jbcdat.rda"))
```

How might you use these data to simulate gene expression profiles for 6 independent and identically distributed samples? (n=20,000 genes)

Get mean and standard deviations for 20,000 features from this data set. 

```{r }


```

Show the distributions of the parameter estimates.  

```{r }  


```

Are the parameter estimates independent?

```{r }  
  

```


Now let's go back and simulate data using these values as our parameters (means and sds). 

1 observation has 20000 genes ~ N(mean_vec, sd_vec), where mean_vec and sd_vec are the vectors of 20000 means and 20000 standard deviations we computed from our data set. 

We want to simulate data under the null, so let's simulate 6 independent observations.  

IMPORTANT:  Make sure everyone in your group simulates the same data.

```{r }
# rnorm(n, mean = 0, sd = 1)


```

Let's compute the averages and sds of our simulated data in the 2 treatment groups.
```{r }
gp1_avg <- 
gp2_avg <-

gp1_sd <-
gp2_sd <-
```

What do you expect to see when you plot gp1_avg vs gp2_avg?

Let's check:
```{r }  
  

```

What do you expect to see when you plot gp1_sd vs gp2_sd?

Let's check:
```{r }  
  

```


What do you see?

Let's plot gp1_sd (estimate) vs sd_vec (truth) and gp2 (estimate) vs sd_vec (truth).

```{r s}  




```

What do you see?

What are the implications for the t-test?

```{r ttests}
tt=genefilter::rowttests(simdat,
                         factor(rep(c("gp1","gp2"),each = 3)))
tt <- cbind.data.frame(tt,
                        pooled_se = tt$dm/tt$statistic)
```

```{r plot_tstat_vs_pooledse}
smoothScatter(tt$pooled_se,tt$statistic)
abline(qt(0.025,4),0,col=2,lwd=2)
abline(qt(0.975,4),0,col=2,lwd=2)
```

```{r plot_tstat_vs_pooledse2}
tt <- 
tt %>% 
  mutate(poolse_decile = ntile(pooled_se, 10))
  
sumquant <- 
  tt %>%
  group_by(poolse_decile) %>%
  summarize(emp.pval_05 = mean(p.value<0.05),
            emp.pval_01 = mean(p.value<0.01))
  sumquant <- cbind.data.frame(sumquant,
      quantpldse = quantile(tt$pooled_se,probs=seq(0.05,0.95,.1)))

plot(sumquant$quantpldse,sumquant$emp.pval_05,
     type="l", 
     xlab = "Pooled SD", 
     ylab = "Pr(|T| > t)",
     main = "False-positive rate")
text(0.11,0.15,"\U03B1 = 0.05")
lines(sumquant$quantpldse,sumquant$emp.pval_01,
      lty=2)
text(0.12,.025,"\U03B1 = 0.1")
```

What does this show us about the results of the t-test under the null hypothesis?

The moderated t-test will address this.